#1644 SyncLab ArtisticEffect Fails to Sync

alexfjw opened the pull request:
Fixes #1644 

### **Outline of Solution**

Since 2013 permanently applies ArtisticEffect to shapes after powerpoint in `SyncLabShapeStorage.cs` is saved, do the following:
Only for 2013, 
1) stash all ArtisticEffects when adding new shapes to SyncLabShapeStorage
2) load these ArtisticEffects back to the shape when retrieving shapes from SyncLabShapeStorage
3) clear the ArtisticEffects as appropriate, when shapes are deleted from SyncLabShapeStorage

<!-- Mention things you want reviewers to focus on, if any. -->
This fix should not hinder ArtisticEffect syncing on 2010 & 2016.

### Others
PictureEffect was renamed to ArtisticEffect, to reflect the naming in the PowerPoint application, rather than VSTO function name.

The ArtisticEffect test has also been fixed. 
The failing test was due to a mistake in slide numbering.
<!-- Skip this section if fix is trivial. -->

leeyh20 commented on PowerPointLabs/PowerPointLabs/SyncLab/SyncLabShapeStorage.cs:
```diff
@@ -61,8 +62,18 @@ public string CopyShape(Shape shape, Format[] formats)
+                    Shape duplicate = shape.Duplicate()[1];
+                    ArtisticEffectFormat.RemoveArtisticEffect(duplicate);
+                    
+                    duplicate.Copy();
```
This will affect the clipboard right? Do you want to use the RestoreClipboardfAterAction function?

leeyh20 commented on PowerPointLabs/Test/UnitTest/SyncLab/SyncLabArtisticEffectTest.cs:
```diff
@@ -4,48 +4,47 @@
         [TestMethod]
         [TestCategory("UT")]
-        public void TestReplaceArtisticEffect()
+        public void TestSetArtisticEffect()
```
Would it be better to name it TestSyncArtisticEffect so that it is consistent with the other synclab tests?

alexfjw replied:
There's another method called TestSyncArtisticEffect, which does that.
ArtisticEffect is a little different from other formats as we have to replace all existing artistic effects on the destination object. This test focuses on the 'replacing' part.

alexfjw commented:
@leeyh20 I've found a bug in my previous approach. 
Have used a different method to ensure ArtisticEffect syncs now.

alexfjw commented on PowerPointLabs/PowerPointLabs/SyncLab/ObjectFormats/ArtisticEffectFormat.cs:
```diff
@@ -0,0 +1,114 @@
+                // access PictureEffects, just to make sure shapes are compatible with ArtisticEffect
+                // will throw an exception otherwise
+                PictureEffects dest = newShape.Fill.PictureEffects;
+                PictureEffects source = formatShape.Fill.PictureEffects;
```
We access the effects just to see if an exception is thrown

IanTeo replied:
Continuing from #1640, discussion on SonarQube.
For example such errors, I remember we had a discussion on this before and concluded that Microsoft suggests to use unused variables to improve readability of the code. Such rules should be removed from SonarQube in that case.

alexfjw commented on PowerPointLabs/PowerPointLabs/SyncLab/ObjectFormats/ArtisticEffectFormat.cs:
```diff
@@ -0,0 +1,114 @@
+            {
+                // access PictureEffects, just to make sure shapes are compatible with ArtisticEffect
+                // will throw an exception otherwise
+                PictureEffects dest = newShape.Fill.PictureEffects;
```
We access the effects just to see if an exception is thrown

leeyh20 commented:
@alexfjw  Oops your syncing of artistic effect does not sync the options?

![image](https://user-images.githubusercontent.com/7023320/38463232-5d83a4aa-3b28-11e8-912a-1ef0687a1787.png)

These special options are not synced. However, every artistic effect has slightly different options:

![image](https://user-images.githubusercontent.com/7023320/38463239-73c3928e-3b28-11e8-9bdc-a5e05a47ade4.png)

alexfjw commented:
Good catch there. It seems that each picture effect might have special parameters to set. 
I didn't notice it as it wasn't very visible in the ArtisticEffect API.
I'll create a new issue to highlight that.

leeyh20 commented:
Okay then! It works on PPT 2010 for the default settings.

leeyh20 commented:
Does it work in 2016?

alexfjw commented:
Yep, it does 👍

IanTeo commented on PowerPointLabs/Test/UnitTest/SyncLab/SyncLabArtisticEffectTest.cs:
```diff
@@ -4,48 +4,47 @@
+            Microsoft.Office.Interop.PowerPoint.Shape newShape = GetShape(sourceSlideNumber, shapeToSync);
+            new ArtisticEffectFormat().SyncFormat(formatShape, newShape);
 
             CompareSlides(sourceSlideNumber, expectedSlideNo);
```
Is this actually enough to compare if the sync actually succeeded? If I remember correctly, this comparison is quite lenient and sometimes we have to make the shape differences very obvious to test it this way. I would suggest we try to test if the attributes directly (if possible) like using IsSameShape, although we would have to add those attributes to be tested too.

alexfjw replied:
I think it's sufficient as the artistic effects we are using for the test greatly changes the look of each shape. I'll look at it once more and check back just to be sure.

alexfjw replied:
Test for Replacing Effects:
![image](https://user-images.githubusercontent.com/11351531/38537015-cfb834ae-3cbe-11e8-9530-c24f2e25ba48.png)

The shape in the top left has it's effects changed after syncing.


Test for Setting Effects:
![image](https://user-images.githubusercontent.com/11351531/38537074-244643da-3cbf-11e8-8c7b-19a1179ef99b.png)

We set a checkered effect here. It also causes the overall lighting of the shape to change as well. 

I believe the change in images is large enough to be sufficient when we compare the slides.

alexfjw replied:
Also note that ArtisticEffect is only applicable to picture type shapes. I'll be adding this in as a comment as well.

IanTeo commented on PowerPointLabs/PowerPointLabs/SyncLab/ObjectFormats/ArtisticEffectFormat.cs:
```diff
@@ -0,0 +1,114 @@
+            List<MsoPictureEffectType> artisticEffects = new List<MsoPictureEffectType>();
+            try
+            {
+                PictureEffects effects = shape.Fill.PictureEffects;
```
Is it possible to choose multiple artistic effects? How do you do it (I am using ppt 2013). If it is possible, then our tests are a bit lacking, we should have a test for this case.

alexfjw replied:
I think you can do so by adding one artistic effect after another. Selecting a new effect adds a new one, rather than replaces the old one. There's a 'clear' button which clears all artistic effects.

IanTeo replied:
So are you going to add tests for such a case? Or is it already covered by one of the tests?

alexfjw replied:
Deepest apologies Ian I've made a mistake on my statement that we can apply more than 1 artistic effect on picture, 
It appears that we can't actually stack artistic effects despite how it sounds in the API. It does however, make me wonder why the API treats ArtisticEffect as a collection of effects. 
I'll note this down as a comment in the ArtisticEffectFormat file.

IanTeo commented on PowerPointLabs/PowerPointLabs/SyncLab/ObjectFormats/ArtisticEffectFormat.cs:
```diff
@@ -0,0 +1,114 @@
+                PictureEffects source = formatShape.Fill.PictureEffects;
+                
+                List<MsoPictureEffectType> effectTypes = GetArtisticEffects(formatShape);
+                ClearArtisticEffects(newShape);
```
Why do we have the effects of the new shape? Is it because it will not override the artistic effect if there is already one?

alexfjw replied:
That's right. I'll add that as a comment to the file to make it clearer.

alexfjw commented:
Thanks for the clarification on slide similarity Ian, 
I did not expect the slide similarity test to fail, but it did as you suspected.
I've changed the test to work on a more obvious artistic effect. (see screenshot below)

The 2 slides compared are the following, which gives a base difference of 96% without syncing.
I've bumped up the slide similarity threshold to 99% for these tests. 
Also, I've added some comments to the slide similarity test in the base SyncLab test file to highlight this.

![image](https://user-images.githubusercontent.com/11351531/38607458-40f020c2-3dab-11e8-994d-2694006901a8.png)
